packages/pigeon/lib/objc_generator.dart

```diff
@@ -1104,26 +1094,42 @@ static id GetNullableObjectAtIndex(NSArray *array, NSInteger key) {
           indent.addln('];');
         });
       });
-      indent.write('[channel sendMessage:$sendArgument reply:^(id reply) ');
+      final String valueOnErrorResponse = func.returnType.isVoid ? '' : 'nil, ';
+      indent.write(
+          '[channel sendMessage:$sendArgument reply:^(NSArray<id> *reply) ');
       indent.addScoped('{', '}];', () {
-        if (func.returnType.isVoid) {
-          indent.writeln('completion(nil);');
-        } else {
-          if (isEnum(root, func.returnType)) {
-            if (func.returnType.isNullable) {
-              indent.writeln(
-                  'NSNumber *outputAsNumber = reply == [NSNull null] ? nil : reply;');
+        indent.writeScoped('if (reply != nil) {', '} ', () {
+          indent.writeScoped('if (reply.count > 1) {', '} ', () {
+            indent.writeln(
+                'completion($valueOnErrorResponse[FlutterError errorWithCode:reply[0] message:reply[1] details:reply[2]]);');
+          }, addTrailingNewline: false);
+          if (!func.returnType.isNullable && !func.returnType.isVoid) {
+            indent.addScoped('else if (reply[0] == nil) {', '} ', () {
               indent.writeln(
-                  '${_enumName(returnType.baseName, suffix: ' *', prefix: languageOptions.prefix, box: true)}output = outputAsNumber == nil ? nil : [[${_enumName(returnType.baseName, prefix: languageOptions.prefix, box: true)} alloc] initWithValue: [outputAsNumber integerValue]];');
+                  'completion($valueOnErrorResponse[FlutterError errorWithCode:@"null-error" message:@"Flutter api returned null value for non-null return value." details:@""]);');
+            }, addTrailingNewline: false);
+          }
+          indent.addScoped('else {', '}', () {
+            const String nullCheck =
+                'reply[0] == [NSNull null] ? nil : reply[0]';
+            if (func.returnType.isVoid) {
+              indent.writeln('completion(nil);');
             } else {
-              indent.writeln(
-                  '${returnType.baseName} output = [reply integerValue];');
+              if (isEnum(root, func.returnType)) {
+                indent.writeln('NSNumber *outputAsNumber = $nullCheck;');
+                indent.writeln(
+                    '${_enumName(returnType.baseName, suffix: ' *', prefix: languageOptions.prefix, box: true)}output = outputAsNumber == nil ? nil : [[${_enumName(returnType.baseName, prefix: languageOptions.prefix, box: true)} alloc] initWithValue: [outputAsNumber integerValue]];');
```

Comment:
This line is a lot; maybe extract the enum name to a local? That would also let you re-use it, and add the * in the string, rather than call twice.